### PR TITLE
refactor: Consolidate `WorkflowService.getMany()` (no-changelog)

### PR DESCRIPTION
**Pull request by ivov**
In scope:

- Consolidate `WorkflowService.getMany()`.
- Support non-entity field `ownedBy` for `select`.
- Support `tags` for `filter`.
- Move `addOwnerId` to `OwnershipService`.
- Remove unneeded check for `filter.id`.
- Simplify DTO validation for `filter` and `select`.
- Expand tests for `GET /workflows`.

Workflow list query DTOs:

```
filter → name, active, tags
select → id, name, active, tags, createdAt, updatedAt, versionId, ownedBy
```

Out of scope:

- Migrate `shared_workflow.roleId` and `shared_credential.roleId` to string IDs.
- Refactor `WorkflowHelpers.getSharedWorkflowIds()`.

### CODE REVIEW

File: packages/cli/src/middlewares/listQuery/dtos/workflow.filter.dto.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	name: string;
+
+	@IsBoolean()
+	@IsOptional()
+	active?: boolean;
+
+	@IsArray()
+	@IsString({ each: true })
+	@IsOptional()
+	tags?: string[];
```
Comment (krynble) [inline packages/cli/src/middlewares/listQuery/dtos/workflow.filter.dto.ts:17]: Amazing code! Filtering by tags doesn't seem to work tho. Could you please check?

Everything else is just perfect 💯

**Review by krynble — CHANGES_REQUESTED** [review]

File: packages/cli/src/middlewares/listQuery/dtos/workflow.filter.dto.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+		Object.assign(this, onlyKnownFields);
+	}
+
+	async validate() {
+		const result = await validate(this);
+
+		if (result.length > 0) throw new Error('Parsed filter does not fit the schema');
+
+		return instanceToPlain(this);
```
Comment (netroy) [inline packages/cli/src/middlewares/listQuery/dtos/workflow.filter.dto.ts:38]: why is validate returning a plain object?

File: packages/cli/src/middlewares/listQuery/dtos/workflow.filter.dto.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	constructor(dto: unknown) {
+		if (!isObjectLiteral(dto)) throw new Error('Filter must be an object literal');
+
+		const onlyKnownFields = Object.fromEntries(
+			Object.entries(dto).filter(([key]) => WorkflowFilter.filterableFields.has(key)),
+		);
+
+		Object.assign(this, onlyKnownFields);
+	}
+
```
Comment (netroy) [inline packages/cli/src/middlewares/listQuery/dtos/workflow.filter.dto.ts:32]: could this be accomplished using `plainToClass`/`plainToInstance` ?

**Review by netroy — COMMENTED** [review]
haven't tested, but code looks good so far.

Once we have more `query` and `body` data as DTO classes, we can setup a decorator and a middleware to automatically transform and validate the inputs, so that we don't have repeat that code in multiple places.

👍🏽

File: packages/cli/src/middlewares/listQuery/dtos/workflow.filter.dto.ts (hunk 1 above)
Comment (ivov) [reply packages/cli/src/middlewares/listQuery/dtos/workflow.filter.dto.ts:17]: Thanks :)

Can you provide more details? I see it working [in Insomnia](https://share.cleanshot.com/H282wS6D) and the [workflow filter test](https://github.com/n8n-io/n8n/pull/6892/files#diff-e5c921d2cb1e160cd0a1d479024a9194bee51289db61cbcd3b834e6d6be3c9e9R168) is passing.

**Review by krynble — DISMISSED** [review]
nvm it all works fine.

I was testing with a single workflow matching a given tag but forgot to remove the `skip=5` param

File: packages/cli/src/middlewares/listQuery/dtos/workflow.filter.dto.ts (hunk 2 above)
Comment (ivov) [reply packages/cli/src/middlewares/listQuery/dtos/workflow.filter.dto.ts]: Please see [`60b4990`](https://github.com/n8n-io/n8n/pull/6892/commits/60b49908f9723408ff3f2b9e1d1dc83b0cfc3c31).

File: packages/cli/src/middlewares/listQuery/dtos/workflow.filter.dto.ts (hunk 3 above)
Comment (ivov) [reply packages/cli/src/middlewares/listQuery/dtos/workflow.filter.dto.ts]: Please see [`60b4990`](https://github.com/n8n-io/n8n/pull/6892/commits/60b49908f9723408ff3f2b9e1d1dc83b0cfc3c31).

**Review by krynble — DISMISSED** [review]

File: packages/cli/src/databases/entities/SharedCredentials.ts [hunk 4]
```diff
@@ -10,7 +10,7 @@ export class SharedCredentials extends WithTimestamps {
 	role: Role;
 
 	@Column()
-	roleId: string;
+	roleId: string; // @TODO: This should be string but is currently number
```
Comment (netroy) [inline packages/cli/src/databases/entities/SharedCredentials.ts:13]: are we planning to change this column type? if not, maybe we should remove this `TODO`.

Comment (ivov) [reply packages/cli/src/databases/entities/SharedCredentials.ts:13]: It's something I noticed but out of scope here. Or you mean we should do it in this PR?

Comment (netroy) [reply packages/cli/src/databases/entities/SharedCredentials.ts:13]: I'm just wondering if there is any plan to change `roleId` to `string` in the DB, and if there are any valid reasons to do so? 
also, we might be able to drop this field all together, and use `role.id`. 
either way, I personally don't find value in this `TODO` comment. 

maybe @krynble has an opinion.

Comment (ivov) [reply packages/cli/src/databases/entities/SharedCredentials.ts]: Synced with Omar, we'll remove the comments and add a transformer to stringify the ID as a fast follow. The original motivation for this was that I found [this check without `toString()` wasn't working](https://github.com/n8n-io/n8n/pull/6892/files#diff-9605a17e2cb66487c428992ac386a0b09df467ac6ef84ef146f3d04fb3eabc24R46).

File: packages/cli/src/databases/migrations/common/1659888469333-AddJsonKeyPinData.ts [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import { isObjectLiteral } from '@/utils';
 import type { IDataObject, INodeExecutionData } from 'n8n-workflow';
 import type { MigrationContext, IrreversibleMigration } from '@db/types';
+import { isObjectLiteral } from '@/utils';
```
Comment (netroy) [inline packages/cli/src/databases/migrations/common/1659888469333-AddJsonKeyPinData.ts:4]: looks like a bad merge. the import already exists on line 1

**Review by flipswitchingmonkey — APPROVED** [review]

File: packages/cli/src/middlewares/listQuery/dtos/pagination.dto.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+	static fromString(rawTake: string, rawSkip: string) {
+		if (!isIntegerString(rawTake)) {
+			throw new Error('Parameter take is not an integer string');
+		}
+
+		if (!isIntegerString(rawSkip)) {
+			throw new Error('Parameter skip is not an integer string');
+		}
+
+		const [take, skip] = [rawTake, rawSkip].map((o) => parseInt(o, 10));
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/cli/src/middlewares/listQuery/dtos/pagination.dto.ts:13]: the `isIntegerString` checks above should prevent that, right?

**Comment by janober** [comment]
Got released with [`n8n@1.4.1`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.4.1)

**Pull request merged**